File: ts/packages/luzid-grpc-client/src/grpc-client.ts
import { AppClient } from './client/app'
import { MutatorClient } from './client/mutator'
import { RpcClient } from './client/rpc'
import { SnapshotClient } from './client/snapshot'
import { StoreClient } from './client/store'
import { ValidatorClient } from './client/validator'
import { Channel, createChannel } from '@luzid/grpc-connection'
import { TransactionClient } from './client/transaction'
import { WorkspaceClient } from './client/workspace'
import { PingClient } from './client/ping'
import { MetaClient } from './client/meta'

const DEFAULT_GRPC_SERVER_PORT = 60061
const DEFAULT_GRPC_SERVER_HOST = 'localhost'

export type LuzidGrpcClientOpts = {
  host?: string
  port?: number
}

export class LuzidGrpcClient {
  private readonly channel: Channel
  private _app?: AppClient
  private _ping?: PingClient
  private _meta?: MetaClient
  private _mutator?: MutatorClient
  private _rpc?: RpcClient
  private _snapshot?: SnapshotClient
  private _store?: StoreClient
  private _validator?: ValidatorClient
  private _transaction?: TransactionClient
  private _workspace?: WorkspaceClient

  constructor(opts?: LuzidGrpcClientOpts) {
    const host = opts?.host ?? DEFAULT_GRPC_SERVER_HOST
    const port = opts?.port ?? DEFAULT_GRPC_SERVER_PORT
    const url = `${host}:${port}`
    this.channel = createChannel(url)
  }

  /**
   * Provides access to the Luzid App service with the following methods:
   *
   * - **appOps**: Controls the Luzid App, i.e to shut it down.
   */
  get app() {
    if (this._app == null) {
      this._app = new AppClient(this.channel)
    }
    return this._app
  }

  /**
   * Provides access to the Luzid Meta service with the following methods:
   *
   * - **getMeta**: Returns the meta information for the Luzid backend.
   */
  get meta() {
    if (this._meta == null) {
      this._meta = new MetaClient(this.channel)
    }
    return this._meta
  }

  /**
   * Provides access to the Luzid Mutator service with the following methods:
   *
   * - **cloneAccount**: Clones an account.
   */
  get mutator() {
    if (this._mutator == null) {
      this._mutator = new MutatorClient(this.channel)
    }
    return this._mutator
  }

  /**
   * Provides access to the Ping client which is used to ensure that the Luzid
   * backend is up via the following methods:
   *
   * - **ping**: Online backend will reply with `{ pong: true }`
   */
  get ping() {
    if (this._ping == null) {
      this._ping = new PingClient(this.channel)
    }
    return this._ping
  }

  /**
   * Provides access to the Luzid RPC service with the following methods:
   *
   * - **getAccountInfo**: Returns the info for an account.
   *  - **requestAirdrop**: Requests to drop SOL to an account.
   */
  get rpc() {
    if (this._rpc == null) {
      this._rpc = new RpcClient(this.channel)
    }
    return this._rpc
  }

  /**
   * Provides access to the Luzid Snapshot service with the following methods:
   *
   * - **getSnaphotableAccounts**: Returns a list of accounts that can be snapshotted.
   * - **listSnapshots**: Returns a list of snapshots created previously.
   * - **retrieveAccountsInSnapshot**: Returns a list of accounts in a snapshot.
   * - **restoreAccountsFromSnapshot**: Restores specific accounts from a snapshot.
   */
  get snapshot() {
    if (this._snapshot == null) {
      this._snapshot = new SnapshotClient(this.channel)
    }
    return this._snapshot
  }

  /**
   * Provides access to the Luzid Store service with the following methods:
   *
   * - **getAccountData**: Returns the data for an account.
   * - **getDiffedAccountUpdate**: Returns the diffed account update for an account related to a transaction.
   */
  get store() {
    if (this._store == null) {
      this._store = new StoreClient(this.channel)
    }
    return this._store
  }

  /**
   * Provides access to the Luzid Validator service with the following methods:
   *
   * - **validatorOps**: Allows to Start/Stop/Restart the validator.
   */
  get validator() {
    if (this._validator == null) {
      this._validator = new ValidatorClient(this.channel)
    }
    return this._validator
  }

  /**
   * Provides access to the Luzid Transaction service with the following methods:
   *
   * - **labelTransaction**: Labels a transaction
   */
  get transaction() {
    if (this._transaction == null) {
      this._transaction = new TransactionClient(this.channel)
    }
    return this._transaction
  }

  /**
   * Provides access to the Luzid Workspace service with the following methods:
   *
   * - **getWorkspace**: Returns the workspace for a given workspace root.
   * - **cloneWorkspace**: Clones a workspace.
   * - **watchWorkspace**: Watches a workspace.
   * - **unwatchWorkspace**: Unwatches a workspace.
   * - **addWorkspace**: Adds a workspace.
   * - **removeWorkspace**: Removes a workspace.
   */
  get workspace() {
    if (this._workspace == null) {
      this._workspace = new WorkspaceClient(this.channel)
    }
    return this._workspace
  }
}

File: dart/luzid-grpc-client/lib/src/grpc_client.dart
import 'dart:async';

import 'package:grpc/grpc.dart';
import 'package:grpc/grpc_connection_interface.dart';
import 'package:luzid_grpc_client/src/client/meta.dart';
import 'package:luzid_grpc_client/src/client/workspace.dart';
import 'package:luzid_grpc_client/src/core/channel.dart';

import 'client/app.dart';
import 'client/mutator.dart';
import 'client/ping.dart';
import 'client/release_info.dart';
import 'client/rpc.dart';
import 'client/snapshot.dart';
import 'client/store.dart';
import 'client/transaction.dart';
import 'client/validator.dart';

export 'client/app.dart';
export 'client/meta.dart';
export 'client/mutator.dart';
export 'client/ping.dart';
export 'client/release_info.dart';
export 'client/rpc.dart';
export 'client/snapshot.dart';
export 'client/store.dart';
export 'client/transaction.dart';
export 'client/validator.dart';

export 'package:grpc/grpc_connection_interface.dart' show ClientChannelBase;

const defaultHost = 'localhost';
const defaultPort = 50051;

class LuzidGrpcClientOpts {
  String? host;
  int? port;

  LuzidGrpcClientOpts({this.host, this.port});
}

class LuzidGrpcClient {
  final ClientChannelBase _channel;
  final String host;
  final int port;
  bool _isShutdown;
  AppClient? _app;
  MetaClient? _meta;
  MutatorClient? _mutator;
  PingClient? _ping;
  RpcClient? _rpc;
  StoreClient? _store;
  SnapshotClient? _snapshot;
  TransactionClient? _transaction;
  ValidatorClient? _validator;
  WorkspaceClient? _workspace;
  final StreamController<bool> _channelConnected = StreamController.broadcast();

  LuzidGrpcClient({LuzidGrpcClientOpts? opts, ClientChannelBase? channel})
      : host = opts?.host ?? defaultHost,
        port = opts?.port ?? defaultPort,
        _isShutdown = false,
        _channel = channel ??
            createLuzidGrpcChannel(
                host: opts?.host ?? defaultHost,
                port: opts?.port ?? defaultPort) {
    bool connected = false;
    _channel.onConnectionStateChanged.listen((state) {
      switch (state) {
        case ConnectionState.ready:
          if (!connected) {
            connected = true;
            print('Connected to gRPC server at port ${opts?.port}');
            _channelConnected.add(connected);
          }
          break;
        case ConnectionState.shutdown:
          if (!_isShutdown && connected) {
            connected = false;
            _channelConnected.add(connected);
          }
          _isShutdown = true;
          break;
        default:
          if (connected) {
            connected = false;
            _channelConnected.add(connected);
          }
          break;
      }
    });
  }

  AppClient get app {
    _app ??= AppClient(_channel);
    return _app!;
  }

  MetaClient get meta {
    _meta ??= MetaClient(_channel);
    return _meta!;
  }

  MutatorClient get mutator {
    _mutator ??= MutatorClient(_channel);
    return _mutator!;
  }

  PingClient get ping {
    _ping ??= PingClient(_channel);
    return _ping!;
  }

  ReleaseInfoClient get releaseInfo {
    return ReleaseInfoClient(_channel);
  }

  RpcClient get rpc {
    _rpc ??= RpcClient(_channel);
    return _rpc!;
  }

  SnapshotClient get snapshot {
    _snapshot ??= SnapshotClient(_channel);
    return _snapshot!;
  }

  StoreClient get store {
    _store ??= StoreClient(_channel);
    return _store!;
  }

  TransactionClient get transaction {
    _transaction ??= TransactionClient(_channel);
    return _transaction!;
  }

  ValidatorClient get validator {
    _validator ??= ValidatorClient(_channel);
    return _validator!;
  }

  WorkspaceClient get workspace {
    _workspace ??= WorkspaceClient(_channel);
    return _workspace!;
  }

  Future<void> close() {
    _isShutdown = true;
    return _channel.shutdown();
  }

  bool get isShutdown {
    return _isShutdown;
  }

  Stream<bool> get onChannelConnected =>
      _channelConnected.stream.where((event) => event);
  Stream<bool> get onChannelDisconnected =>
      _channelConnected.stream.where((event) => !event);
}
